File: provision/lib/platform.lua
-- provision/lib/platform.lua
--
-- Platform detection and helpers for the Lua provisioner.
-- Platform is read from the first CLI argument passed to main.lua ("mac" or
-- "windows"/"win"). Any other value causes an immediate error exit.

local M = {}

-- Allowlist of recognized platform strings → canonical name
local PLATFORM_ALIASES = {
  mac     = "mac",
  windows = "windows",
  win     = "windows",
}

-- Resolved platform, set by M.init()
local _platform = nil

--- Initialize platform from the CLI argument.
-- Must be called once at startup (main.lua). Exits with an error message if
-- the platform argument is missing or not in the allowlist.
-- @param arg string  The first CLI argument (arg[1]).
function M.init(arg_platform)
  if not arg_platform or arg_platform == "" then
    io.stderr:write("[platform] ERROR: platform argument is required (mac | windows)\n")
    os.exit(2)
  end

  local canonical = PLATFORM_ALIASES[arg_platform:lower()]
  if not canonical then
    io.stderr:write(string.format(
      "[platform] ERROR: unsupported platform %q — must be one of: mac, windows\n",
      arg_platform
    ))
    os.exit(2)
  end

  _platform = canonical
end

--- Return the canonical platform name ("mac" or "windows").
function M.current()
  assert(_platform, "platform.init() has not been called")
  return _platform
end

--- Return true when running on macOS.
function M.is_mac()
  return M.current() == "mac"
end

--- Return true when running on Windows.
function M.is_windows()
  return M.current() == "windows"
end

--- Return the user home directory.
-- Uses $HOME on mac/linux-style systems; $USERPROFILE on Windows.
function M.home()
  return os.getenv("HOME") or os.getenv("USERPROFILE") or ""
end

--- Return the dotfiles root directory.
-- Honors $DF_ROOT_DIRECTORY if set; otherwise defaults to
-- $HOME/Documents/GitHub/dotfiles.
function M.dotfiles_root()
  return os.getenv("DF_ROOT_DIRECTORY") or (M.home() .. "/Documents/GitHub/dotfiles")
end

--- Return the data directory used for state persistence (~/.df_data).
function M.data_dir()
  return M.home() .. "/.df_data"
end

--- Create a directory (and any missing parents) idempotently.
-- On mac uses `mkdir -p`. On Windows shells `mkdir -p` would create a literal
-- directory named "-p", so we use PowerShell's `New-Item -Force` which is the
-- native equivalent.
-- @param path string  Filesystem path to create.
-- @return boolean     true on success.
function M.mkdir_p(path)
  if M.is_windows() then
    local cmd = string.format(
      'powershell -NoProfile -Command "New-Item -ItemType Directory -Force -Path \'%s\' | Out-Null"',
      path:gsub("'", "''")
    )
    return os.execute(cmd)
  end
  return os.execute(string.format('mkdir -p %q', path))
end

--- Return true if an entry's platform list includes the current platform.
-- If entry.platforms is nil or empty, the entry applies to all platforms.
-- @param entry table  A manifest entry (tool/script/configurator).
function M.matches(entry)
  if not entry.platforms or #entry.platforms == 0 then
    return true
  end
  local cur = M.current()
  for _, p in ipairs(entry.platforms) do
    if PLATFORM_ALIASES[p:lower()] == cur then
      return true
    end
  end
  return false
end

return M

File: provision/lib/platform.ts
// provision/lib/platform.ts
//
// Platform detection and helpers for the Bun provisioner.
// Platform is read from the first CLI argument passed to main.ts ("mac" or
// "windows"/"win"). Any other value causes an immediate error exit.
import {mkdirSync} from "node:fs";
import {homedir} from "node:os";

export type Platform = "mac" | "windows";

// Resolved platform, set by init().
let _platform: Platform | null = null;

// Allowlist of recognized platform strings → canonical name.
const PLATFORM_ALIASES: Record<string, Platform> = {
    mac: "mac",
    windows: "windows",
    win: "windows",
};

/** Return the canonical platform name ("mac" or "windows"). */
export function current(): Platform {
    if (!_platform) {
        throw new Error("platform.init() has not been called");
    }

    return _platform;
}

/**
 * Return the user home directory.
 * Uses $HOME on mac/linux-style systems; $USERPROFILE on Windows.
 */
export function home(): string {
    return process.env.HOME || process.env.USERPROFILE || homedir() || "";
}

/** Return the data directory used for state persistence (~/.df_data). */
export function dataDir(): string {
    return `${home()}/.df_data`;
}

/**
 * Return the dotfiles root directory.
 * Honors $DF_ROOT_DIRECTORY if set; otherwise defaults to
 * $HOME/Documents/GitHub/dotfiles.
 */
export function dotfilesRoot(): string {
    return process.env.DF_ROOT_DIRECTORY || `${home()}/Documents/GitHub/dotfiles`;
}

/**
 * Initialize platform from the CLI argument.
 * Must be called once at startup (main.ts). Exits with an error message if
 * the platform argument is missing or not in the allowlist.
 */
export function init(argPlatform: string | undefined): void {
    if (!argPlatform || argPlatform === "") {
        process.stderr.write("[platform] ERROR: platform argument is required (mac | windows)\n");
        process.exit(2);
    }

    const canonical = PLATFORM_ALIASES[argPlatform.toLowerCase()];
    if (!canonical) {
        process.stderr.write(
            `[platform] ERROR: unsupported platform "${argPlatform}" — must be one of: mac, windows\n`,
        );

        process.exit(2);
    }

    _platform = canonical;
}

/** Return true when running on macOS. */
export function isMac(): boolean {
    return current() === "mac";
}

/** Return true when running on Windows. */
export function isWindows(): boolean {
    return current() === "windows";
}

/**
 * Return true if an entry's platform list includes the current platform.
 * If entry.platforms is nil or empty, the entry applies to all platforms.
 */
export function matches(entry: { platforms?: string[] }): boolean {
    if (!entry.platforms || entry.platforms.length === 0) {
        return true;
    }

    const cur = current();
    return entry.platforms.some((p) => PLATFORM_ALIASES[p.toLowerCase()] === cur);
}

/**
 * Create a directory (and any missing parents) idempotently.
 * Native recursive mkdir — no shell involved, so it is portable across
 * mac and Windows without the `mkdir -p` / literal `-p` directory pitfall.
 */
export function mkdirP(path: string): void {
    mkdirSync(path, {recursive: true});
}